server/src/main/java/org/opensearch/cluster/routing/RoutingTableIncrementalDiff.java

```diff
@@ -0,0 +1,173 @@
+        this.diffs = diffs;
+    }
+
+    /**
+     * Gets the map of differences of {@link IndexRoutingTable}.
+     *
+     * @return a map containing the differences.
+     */
+    public Map<String, Diff<IndexRoutingTable>> getDiffs() {
+        return diffs;
+    }
+
+    /**
+     * Reads a {@link RoutingTableIncrementalDiff} from the given {@link StreamInput}.
+     *
+     * @param in the input stream to read from.
+     * @return the deserialized RoutingTableIncrementalDiff.
+     * @throws IOException if an I/O exception occurs while reading from the stream.
+     */
+    public static RoutingTableIncrementalDiff readFrom(StreamInput in) throws IOException {
+        int size = in.readVInt();
+        Map<String, Diff<IndexRoutingTable>> diffs = new HashMap<>();
+
+        for (int i = 0; i < size; i++) {
+            String key = in.readString();
+            List<IndexShardRoutingTable> shardRoutingTables = new ArrayList<>();
+
+            // Read each IndexShardRoutingTable from the stream
+            int numShards = in.readVInt();
+            for (int j = 0; j < numShards; j++) {
+                IndexShardRoutingTable shardRoutingTable = IndexShardRoutingTable.Builder.readFrom(in);
+                shardRoutingTables.add(shardRoutingTable);
+            }
+
+            // Create a diff object for the index
+            Diff<IndexRoutingTable> diff = new IndexRoutingTableIncrementalDiff(shardRoutingTables);
+
+            // Put the diff into the map with the key
+            diffs.put(key, diff);
+        }
+        return new RoutingTableIncrementalDiff(diffs);
+    }
+
+    /**
+     * Applies the differences to the provided {@link IndexRoutingTable}.
+     *
+     * @param part the original IndexRoutingTable to which the differences will be applied.
+     * @return the updated IndexRoutingTable with the applied differences.
+     */
+    @Override
+    public IndexRoutingTable apply(IndexRoutingTable part) {
+        // Apply diffs to the provided IndexRoutingTable
+        for (Map.Entry<String, Diff<IndexRoutingTable>> entry : diffs.entrySet()) {
+            part = entry.getValue().apply(part);
+        }
+        return part;
+    }
+
+    /**
+     * Writes the differences to the given {@link StreamOutput}.
+     *
+     * @param out the output stream to write to.
+     * @throws IOException if an I/O exception occurs while writing to the stream.
+     */
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeVInt(diffs.size());
+        for (Map.Entry<String, Diff<IndexRoutingTable>> entry : diffs.entrySet()) {
+            out.writeString(entry.getKey());
+            entry.getValue().writeTo(out);
+        }
+    }
+
+    /**
+     * Represents a difference between {@link IndexShardRoutingTable} objects that can be serialized and deserialized.
+     */
+    public static class IndexRoutingTableIncrementalDiff implements Diff<IndexRoutingTable> {
```

Comment:
Why do we meed `RoutingTableIncrementalDiff`? Cant we directly use this class?

Reply:
We use this code twice for reading and writing, respectively. Therefore, decided to abstract it out for better code reuse.